api/controllers/export-data-2.js

```diff
@@ -0,0 +1,257 @@
+    if (!forms) {
+      p = p.then(() =>
+        DB.query('medic-client/reports_by_form', {
+          group: true
+        }).then(results => results.rows.map(r => r.key[0]))
+      );
+    } else {
+      p = p.then(() => forms);
+    }
+
+    // Determine the fields for each form we care about
+    return p.then(forms =>
+      Promise.all(forms.map(form =>
+        DB.query('medic-client/reports_by_form', {
+          key: [form],
+          limit: 1,
+          include_docs: true,
+          reduce: false
+        })
+        .then(results =>
+          results.rows[0] &&
+          results.rows[0].doc &&
+          results.rows[0].doc.fields
+        ))
+      ).then(allFields => {
+        const fieldColumns = _.union(
+          ...allFields.filter(f => f).map(f => Object.keys(flatten(f)))
+        ).sort();
+
+        const allColumns = [
+          '_id',
+          'patient_id',
+          'reported_date',
+          'from',
+          'contact.name',
+          'contact.parent.name',
+          'contact.parent.parent.name',
+          'contact.parent.parent.parent.name'
+        ].concat(fieldColumns);
+
+        return {
+          header: allColumns,
+          fn: record => {
+            const flattened = flatten(record.fields);
+
+            return [
+              record._id,
+              record.patient_id,
+              record.reported_date,
+              record.from,
+              safeGet(record, ['contact', 'name']),
+              safeGet(record, ['contact', 'parent', 'name']),
+              safeGet(record, ['contact', 'parent', 'parent', 'name']),
+              safeGet(record, ['contact', 'parent', 'parent', 'parent', 'name'])
+            ].concat(fieldColumns.map(c => flattened[c]));
+          }
+        };
+      }));
+  },
+
+
+  // NB: we don't actually support this (note SUPPORTED_EXPORTS above), it's
+  // here in the code to show a simpler examplen than the complicated forms one
+  contacts: () => Promise.resolve({
+    header: ['id', 'rev', 'name', 'patient_id', 'type'],
+    fn: record => [record._id, record._rev, record.name, record.patient_id, record.type]
+  })
+};
+
+class SearchResultReader extends Readable {
+
+  constructor(type, filters, searchOptions, readableOptions) {
+    super(readableOptions);
+
+    this.type = type;
+    this.filters = filters;
+    this.options = searchOptions;
+
+    // There is no reason for a user to pass a skip, but we're going to allow
+    // users to pass a limit. This could be useful as an escape hatch / tweak in
+    // production.
+    this.options.skip = 0;
+    this.options.limit = this.options.limit || BATCH;
+  }
+
+  _read() {
+    if (!this.csvFn) {
+      return Promise.resolve()
+        .then(() => CSV_MAPPER[this.type](this.filters, this.options))
+        .then(({ header, fn }) => {
+          this.csvFn = fn;
+          this.push(header.join(JOIN_COL) + JOIN_ROW);
+        });
+    }
+
+    return Promise.resolve()
```

Comment:
To me it makes it more readable, because things align up in a more consistent way. It's not how the rest of the codebase works though, and if we're worried about ticks (are we?) then that's an issue, so I'll change it to be more like how the rest of the codebase works.